Block legacy get-mac in preference of getmac

mark007 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Remove references to the now deprecated get-mac python library, in favor of the newer getmac library. The getmac library author has recommended users of the old get-mac library move to the new getmac one instead as the old library will cease to receive updates in the near future.

Additionally the two libraries, internally share the same 'getmac' python package, so when installed on the same system, conflict with each other (with get-mac overruling getmac seemingly). This is causing integrations which require fixes in the latest getmac library to fail as get-mac is masking those updates.

get-mac is also added to the package constraints to prevent it from being installed in the future, which may be considered a breaking change for any other integrations not in the core, which still use get-mac as a dependency. They would need to move to use getmac instead in order to be compatible with HA after this is merged.

https://github.com/ManneW/vilfo-api-client-python/compare/v0.3.2...0.4.1

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #87146
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mark007 committed:
Fix for error "Cannot get the ip address of kef speaker"

mark007 commented:
I notice this adds logs like the following, and not just for the kef integration, for all integrations using getmac presumably. I'm thinking now whether this should be set at HA level for this reason.
```

Logger: getmac
Source: /usr/local/lib/python3.10/site-packages/getmac/getmac.py:1615
First occurred: 15:33:31 (4 occurrences)
Last logged: 15:33:43

Forcing method 'ArpFile' to be used for 'ip4' lookup (arg: '192.168.8.162')
Forcing method 'ArpFile' to be used for 'ip4' lookup (arg: '192.168.8.1')
Forcing method 'ArpFile' to be used for 'ip4' lookup (arg: '192.168.8.121')
```

MartinHjelmare commented on homeassistant/components/kef/media_player.py:
```diff
@@ -117,6 +118,7 @@ async def async_setup_platform(
     )
 
     mode = get_ip_mode(host)
+    getmac.getmac.FORCE_METHOD="ArpFile"
```
This should be fixed in the `getmac` library as already discussed here:

https://github.com/GhostofGoes/getmac/issues/83

Regarding the different versions of the library, we can denylist `get-mac` in `package_constraints` like this:

https://github.com/home-assistant/core/blob/1431979eab9cfe158deaba9ccf268f8315abe7a6/homeassistant/package_constraints.txt#L82

Libraries that use `get-mac` should be updated to use `getmac` instead so we eventually can remove the denylist item.

mark007 replied:
You are right. I have asked the dev of one of the ha integrations to update that package in ha to use getmac. I'll add the package constraint in this patch.

mark007 committed:
Update media_player.py

mark007 commented:
I have updated the patch to update the version of vilfo-api-client to 0.4.1 which no longer uses 'get-mac'. I also added a constraint that get-mac can't be installed from now on by other integrations. I would appreciate any help with testing this as locally the constraint seems to work, but I guess its now counted as a breaking change for any other integrations out in the world that might still use get-mac, but are not contained within the homeassistant/core repository?

MartinHjelmare commented on homeassistant/components/vilfo/manifest.json:
```diff
@@ -6,5 +6,5 @@
   "loggers": ["vilfo"],
-  "requirements": ["vilfo-api-client==0.3.2"]
-}
+  "requirements": ["vilfo-api-client==0.4.1"]
```
It would be good if someone can test the vilfo integration with this change.

ManneW replied:
Sorry for the late reply. I'll have a look at this within the next couple of days and if anything needs to be updated in the Vilfo integration, I'll try to do that as well 🙂

MartinHjelmare replied:
Great, thanks!

mark007 replied:
Hi @ManneW just wondering do you think you might have some time to confirm the uplift of the version of your package looks OK to merge.

MartinHjelmare commented on homeassistant/package_constraints.txt:
```diff
@@ -145,3 +145,7 @@ matplotlib==3.6.1
+
+# The get-mac package has been replaced with getmac. Installing get-mac alongside getmac
+# breaks getmac due to them both sharing the same python package name inside 'getmac'.
+get-mac<0.0.0
```
Please run the `gen_requirements_all.py` script to update the requirements.

mark007 replied:
I have run the gen requirements script, but because the get-mac constraint is a negative one, ie never allow get-mac to be included, and now that no core integrations use it anymore, the gen requirements script is not updating the requirements txt files with get-mac, only the vilfo-api-client one.

MartinHjelmare replied:
Try setting it to a high number instead like we do above.

```suggestion
get-mac==1000000000.0.0
```

mark007 replied:
Good idea, done.

MartinHjelmare reviewed (commented):
Please rebase on latest dev branch and solve the merge conflict.

frenck commented:
I've marked this PR, as some jobs in our CI are failing. Please check the output of those jobs to see what went wrong and adjust the PR accordingly.

Please un-draft it once the CI passes and it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

GnatorX commented:
@mark007 It seems like merging in latest broke a test?

GnatorX commented:
If I have to guess, you may need to add extra /n at the end of the file?

GnatorX committed:
Update manifest.json

mark007 commented:
> If I have to guess, you may need to add extra /n at the end of the file?

yes I'll remove that new line to see if it passes. We still require @ManneW to sign off that the uplifted package version is ok from his side though.

MartinHjelmare commented:
Started CI. :+1:

jlpouffier commented:
@mark007 Any news?

mark007 commented:
Unfortunately we haven't been able to get the other required developer to help test the uplift of their integration which is part of the fix. All we can do is wait unfortunately.

jlpouffier commented:
I know you must hear that often from people that do not really mean it but : is three anyway some random dude on the internet (aka. Me) can help?

mark007 commented:
Yeah absolutely if you know how to test the patch in terms of the uplifted viflpo integration, and fix unit tests that seem to fail due to the uplift. I hope the patch could then pass the init tests and the maintainers would allow the patch to merge based on your testing.

jlpouffier commented:
Yeah. I have no clue about what you just said, unfortunately, I won't be able to help :( Sorry...

GnatorX commented:
I will try to take a look over the weekend and see if I can fix the unit tests.

jlpouffier commented:
Hey @GnatorX 
Did you had a chance to look at this during your weekend:) ?
Have a great day.

GnatorX commented:
Still investigating but in test_form (one of the tests that fails) it is throwing an exception during [is_host_valid](https://github.com/home-assistant/core/blob/dev/homeassistant/components/vilfo/config_flow.py#L72). It isn't invalid_host since that should be caught, instead it is throwing up a 'errors': {'base': 'unknown'} which is the catch all. I see that the exception is 
```
2023-03-28 14:39:17.490 ERROR    MainThread homeassistant.components.vilfo.config_flow:config_flow.py:115 Unexpected exception: 'NoneType' object is not subscriptable
```
Will update when I get further

GnatorX commented:
For the test_form test. It is failing in the creation of VilfoClient. I suspect https://github.com/ManneW/vilfo-api-client-python/blob/master/vilfo/client.py#L38.

GnatorX commented:
Specifically it is throwing 

```
'NoneType' object is not subscriptable
```

GnatorX commented:
Ah i think we need to mock calls to _try_connect_and_fetch_basic_info or else that returns none.. Testing

GnatorX commented:
I fixed the tests... setting up a PR against your branch

GnatorX committed:
Update test_config_flow.py

GnatorX committed:
Update manifest.json

GnatorX commented:
https://github.com/mark007/core/pull/1 @mark007

GnatorX committed:
Update test_config_flow.py

mark007 commented:
> [mark007#1](https://github.com/mark007/core/pull/1) @mark007

Thanks a lot for this. I ran the CI flow. Great that the unit tests now pass. There were a few issues with flake8 and black tests. Can you check those, maybe get them passing locally and push another patch. After all is green I'll merge your changes into my patch then and request it be reviewed again.

https://github.com/mark007/core/actions/runs/4548309507/jobs/8026234272

GnatorX committed:
Update test_config_flow.py

GnatorX committed:
Update test_config_flow.py

GnatorX commented:
Sorry, I realize now that we have black and other lint in this repo. I fixed black and flake8 but i think actions are down atm. Waiting for that to recover to re-test https://www.githubstatus.com/

GnatorX commented:
https://github.com/GnatorX/core/actions/runs/4555936517/jobs/8035639482 CI works for black + flake8. I don't think the pip checks is changes I made and should work on your branch? @mark007

GnatorX committed:
Update manifest.json

GnatorX committed:
Update package_constraints.txt

GnatorX committed:
Update requirements_all.txt

GnatorX committed:
Update requirements_test_all.txt

GnatorX committed:
Update gen_requirements_all.py

GnatorX commented:
Fixed the test ( I was targeting your dev branch which was missing the package uplift) https://github.com/GnatorX/core/actions/runs/4556287084/jobs/8036687674

GnatorX commented:
https://github.com/mark007/core/pull/1/checks @mark007 passing ci

mark007 commented:
> Are we ready to move ahead? Please mark the PR as ready for review when ready. Thanks!

I just see a code coverage test failing now. Is that mandatory to pass. If so I guess we need further work. We have not yet had a confirmation from the vilfo integration author that the uplift is fine. Are we still ok to merge without that confirmation.

GnatorX commented:
I am not sure if the coverage test fails because of this change since it failed to upload. @ManneW are you able to test the three vilfo integration? Thanks

GnatorX commented:
@ManneW Hey sorry to ping again, I can't test the integration.. at least not easily since I don't have any vilfo routers. Any chance you can give this a test?

ManneW commented:
> @ManneW Hey sorry to ping again, I can't test the integration.. at least not easily since I don't have any vilfo routers. Any chance you can give this a test?

Sorry for the late reply and lack of activity. Life happened to get in the way. Hopefully, I can have a look tomorrow or Monday.

mark007 commented:
Did you have any luck testing this change @ManneW 
If you were happy to sign off we could request it be merged. It should fix at least two other integrations affected.

mark007 commented:
Hi @ManneW I wonder can you indicate if this is something you can look at soon. If not its no problem, we might need to look at an alternative approach to fix these other integrations, possibly making them use an alternative library to 'getmac'. If you don't have time to have a look from your side, I can start looking into fixing this by using an alternative library in the likes of the 'kef' integration which is broken at the moment without this patch.

breakthestatic commented:
At this point, can we just move ahead with this fix?  Using the analytics data as a base (I know, not perfect...) the affected integration that has not been tested has **four** active installations. The specific integration noted in this ticket, KEF speakers, has *seven** active installations. Additionally (and IMO more importantly), the NMAP tracker integration has **4,465** active installations and is currently broken as of 2023.2+.

GnatorX commented:
Agreed, what is the policy in the home assistant project for moving forward when specific integrations owners are unresponsive?

basnijholt commented:
I am the "owner" of this integration and approved this almost 3 months ago.

Not sure why this has not been merged.

GnatorX commented:
Sorry I meant the Vilfo routers integration which this PR is waiting on for testing/ validation

MartinHjelmare reviewed (approved):
We've decided to merge this PR even though the vilfo changes haven't been locally tested. The current problem with the get-mac package affects other integrations so we think it's best to merge here now.
